File: libs/element/src/el/Widget.lua
--- A GUI Widget
-- @classmod el.Widget
-- @pragma nostrip

local ui        = require ('element.ui')
local object    = require ('kv.object')

local Widget = object {
    --- Widget name (string).
    -- @class field
    -- @name Widget.name
    -- @within Attributes
    name = {
        get = function (self) return self.impl:get_name() end,
        set = function (self, value) self.impl:set_name (value) end
    },

    --- Widget visibility (bool).
    -- @class field
    -- @name Widget.visible
    -- @within Attributes
    visible = {
        get = function (self) return self.impl:is_visible() end,
        set = function (self, value) self.impl:set_visible (value) end
    },

    --- Widget bounding box (table).
    -- The coords returned is relative to the top-left of the widget's parent.
    -- @class field
    -- @name Widget.bounds
    -- @within Attributes
    -- @usage
    -- widget.bounds = {
    --     x      = 0, 
    --     y      = 0,
    --     width  = 100,
    --     height = 200
    -- }
    bounds = {
        get = function (self) return self.impl:get_bounds() end,
        set = function (self, value)
            if 'table' == type(value) then self:set_bounds (value) end
        end
    },

    --- Widget width (integer).
    -- @class field
    -- @name Widget.width
    -- @within Attributes
    width = {
        get = function (self) return self.impl:get_width() end
    },

    --- Widget height (integer).
    -- @class field
    -- @name Widget.height
    -- @within Attributes
    height = {
        get = function (self) return self.impl:get_height() end
    }
}

--- Initialize the widget.
-- Override this to customize your widget
function Widget:init (impl)
    self.impl = impl or ui.ComponentWrapper (self)
end

--- Add a child widget.
-- @tparam el.Widget widget Widget to add
-- @int zorder Z-order (default -1)
function Widget:add (widget, zorder)
    zorder = zorder or -1
    self.impl:add (widget, zorder)
    return widget
end

--- Change this widget's size
-- @int width   New width
-- @int height  New height
function Widget:resize (width, height)
    self.impl:resize (width, height)
end

--- Set the widget's bounds
function Widget:set_bounds (...)
    if select("#", ...) >= 4 then
        self.impl:setbounds (
            select (1, ...),
            select (2, ...),
            select (3, ...),
            select (4, ...)
        )
    else
        local r = select (1, ...)
        self.impl:set_bounds (r.x, r.y, r.width, r.height)
    end
end

--- Called when the widget changes in size.
-- Override this to layout child widget's
function Widget:resized()
end

--- Called when the widget needs drawn
-- @tparam juce.Graphics g Graphics context to draw with
function Widget:paint (g)
end

--- Repaint the entire widget
function Widget:repaint()
    self.impl:repaint()
end

function Widget:mouse_drag (ev) end
function Widget:mouse_down (ev) end
function Widget:mouse_up (ev) end

return Widget

File: libs/element/lua/el/Widget.cpp
/// A GUI Widget.
// Is defined with @{el.object} and can be inherrited. Backed by a JUCE Component.
// @classmod el.Widget
// @pragma nostrip

#include "object.hpp"
#include "widget.hpp"
#define LKV_TYPE_NAME_WIDGET     "Widget"

using namespace juce;

namespace kv {
namespace lua {

class Widget : public juce::Component
{
public:
    ~Widget()
    {
        widget = sol::lua_nil;
    }

    Widget (const sol::table& obj)
    {
        widget = obj;
    }

    void resized() override
    {
        if (sol::safe_function f = widget ["resized"])
            f (widget);
    }

    void paint (Graphics& g) override
    {
        if (sol::safe_function f = widget ["paint"]) {
            f (widget, std::ref<Graphics> (g));
        }
    }

    void mouseDrag (const MouseEvent& ev) override
    {
        if (sol::safe_function f = widget ["mousedrag"])
            f (widget, ev);
    }

    void mouseDown (const MouseEvent& ev) override
    {
        if (sol::safe_function f = widget ["mousedown"])
            f (widget, ev);
    }

    void mouseUp (const MouseEvent& ev) override
    {
        if (sol::safe_function f = widget ["mouseup"])
            f (widget, ev);
    }

    sol::table addWithZ (const sol::object& child, int zorder)
    {
        jassert (child.valid());
        if (Component* const impl = object_userdata<Component> (child))
        {
            addAndMakeVisible (*impl, zorder);
        }

        return child;
    }

    sol::table add (const sol::object& child)
    {
        return addWithZ (child, -1);
    }
    
    static void init (const sol::table& proxy) {
        if (auto* const impl = object_userdata<Widget> (proxy)) {
            impl->widget = proxy;
        }
    }

    sol::table getBoundsTable()
    {
        sol::state_view L (widget.lua_state());
        auto r = getBounds();
        auto t = L.create_table();
        t["x"]      = r.getX();
        t["y"]      = r.getY();
        t["width"]  = r.getWidth();
        t["height"] = r.getHeight();
        return t;
    }

private:
    Widget() = delete;
    sol::table widget;
    JUCE_DECLARE_NON_COPYABLE_WITH_LEAK_DETECTOR (Widget)
};

}}

EL_PLUGIN_EXPORT
int luaopen_el_Widget (lua_State* L) {
    using kv::lua::Widget;

    auto T = kv::lua::new_widgettype<Widget> (L, LKV_TYPE_NAME_WIDGET,
        sol::meta_method::to_string, [](Widget& self) {
            return kv::lua::to_string (self, LKV_TYPE_NAME_WIDGET);
        },
        /// Add a child widget.
        // @function Widget:add
        // @tparam el.Widget widget Widget to add
        // @int[opt] zorder Z-order
        // @within Methods
        "add", sol::overload (&Widget::add, &Widget::addWithZ),
        "addtodesktop", sol::overload (
            [](Widget& self, int flags) { 
                self.addToDesktop(flags, nullptr); 
            },
            [](Widget& self, int flags, void* handle) { 
                self.addToDesktop (flags, handle); 
            }
        ),
        sol::base_classes,      sol::bases<Component>()
    );

    sol::table T_mt = T [sol::metatable_key];
    T_mt["__methods"].get_or_create<sol::table>().add (
        "add"
    );

    sol::stack::push (L, T);

    return 1;
}
